In [6]:
import pandas as pd
import statsmodels.api as sm

from linearmodels.datasets import wage_panel
from linearmodels.panel import PanelOLS
from linearmodels.panel import PooledOLS
from linearmodels.panel import RandomEffects

import scipy.stats as st
import statsmodels.api as sm
import statsmodels.graphics.tsaplots as tsap
from statsmodels.compat import lzip
from statsmodels.stats.diagnostic import het_white
from linearmodels.panel import compare
from sklearn import preprocessing
import datetime as dt

In [7]:
df = pd.read_excel('panel.xlsx')

In [8]:
df

id     t        y       x1     x2  d1  d2  d3  d4  Unnamed: 9  \
0    1  2009  28.2561  17.0188   7.75   0   0   0   0         NaN   
1    1  2010  28.2730  17.0161   8.36   0   0   0   0         NaN   
2    1  2011  28.2801  17.0169   8.36   0   0   0   0         NaN   
3    1  2012  28.2498  17.0395  10.65   0   0   0   0         NaN   
4    1  2013  28.2313  17.0458  12.15   0   0   0   0         NaN   
5    1  2014  28.2312  17.0616  12.68   0   0   0   0         NaN   
6    1  2015  28.2390  17.0602  11.90   0   0   0   0         NaN   
7    1  2016  28.2518  17.0710  11.69   0   0   0   0         NaN   
8    1  2017  28.2684  17.0776  11.21   0   0   0   0         NaN   
9    1  2018  28.2776  17.0799  10.61   0   0   0   0         NaN   
10   1  2019  28.2824  17.0700   9.95   0   0   0   0         NaN   
11   2  2009  28.5855  17.2839   7.54   1   0   0   0         NaN   
12   2  2010  28.6095  17.2886   7.79   1   0   0   0         NaN   
13   2  2011  28.6201  17.2961   8.04   1   0   0   0         NaN   
14   2  2012  28.6345  17.3053   7.88   1   0   0   0         NaN   
15   2  2013  28.6525  17.3139   7.52   1   0   0   0         NaN   
16   2  2014  28.6840  17.3220   6.11   1   0   0   0         NaN   
17   2  2015  28.7077  17.3305   5.30   1   0   0   0         NaN   
18   2  2016  28.7291  17.3401   4.81   1   0   0   0         NaN   
19   2  2017  28.7532  17.3457   4.33   1   0   0   0         NaN   
20   2  2018  28.7701  17.3540   4.00   1   0   0   0         NaN   
21   2  2019  28.7861  17.3621   3.74   1   0   0   0         NaN   
22   3  2009  27.9367  16.7487   8.34   0   1   0   0         NaN   
23   3  2010  27.9671  16.7592   8.06   0   1   0   0         NaN   
24   3  2011  27.9981  16.7676   7.51   0   1   0   0         NaN   
25   3  2012  28.0156  16.7765   7.29   0   1   0   0         NaN   
26   3  2013  28.0386  16.7871   7.07   0   1   0   0         NaN   
27   3  2014  28.0669  16.7904   6.91   0   1   0   0         NaN   
28   3  2015  28.0735  16.7956   6.91   0   1   0   0         NaN   
29   3  2016  28.0834  16.8051   7.00   0   1   0   0         NaN   
30   3  2017  28.1134  16.8185   6.34   0   1   0   0         NaN   
31   3  2018  28.1408  16.8283   5.83   0   1   0   0         NaN   
32   3  2019  28.1594  16.8485   5.66   0   1   0   0         NaN   
33   4  2009  30.4005  18.8742   9.25   0   0   1   0         NaN   
34   4  2010  30.4273  18.8740   9.63   0   0   1   0         NaN   
35   4  2011  30.4426  18.8746   8.95   0   0   1   0         NaN   
36   4  2012  30.4652  18.8810   8.07   0   0   1   0         NaN   
37   4  2013  30.4834  18.8832   7.37   0   0   1   0         NaN   
38   4  2014  30.5061  18.8878   6.17   0   0   1   0         NaN   
39   4  2015  30.5328  18.8947   5.28   0   0   1   0         NaN   
40   4  2016  30.5493  18.9059   4.87   0   0   1   0         NaN   
41   4  2017  30.5715  18.9152   4.36   0   0   1   0         NaN   
42   4  2018  30.6005  18.9233   3.90   0   0   1   0         NaN   
43   4  2019  30.6232  18.9341   3.67   0   0   1   0         NaN   
44   5  2009  28.7176  17.5527   7.74   0   0   0   1         NaN   
45   5  2010  28.7585  17.5520   6.97   0   0   0   1         NaN   
46   5  2011  28.7970  17.5467   5.82   0   0   0   1         NaN   
47   5  2012  28.8012  17.5497   5.38   0   0   0   1         NaN   
48   5  2013  28.8056  17.5582   5.23   0   0   0   1         NaN   
49   5  2014  28.8274  17.5640   4.98   0   0   0   1         NaN   
50   5  2015  28.8422  17.5688   4.62   0   0   0   1         NaN   
51   5  2016  28.8643  17.5898   4.12   0   0   0   1         NaN   
52   5  2017  28.8907  17.5956   3.75   0   0   0   1         NaN   
53   5  2018  28.9005  17.5982   3.38   0   0   0   1         NaN   
54   5  2019  28.9110  17.6095   3.14   0   0   0   1         NaN   

    Unnamed: 10  Unnamed: 11       Unnamed: 12  Unnamed: 13  \
0           NaN          NaN               NaN          NaN   
1           N

In [9]:
df = df.set_index(['t', 'id'])
df

y       x1     x2  d1  d2  d3  d4  Unnamed: 9  Unnamed: 10  \
t    id                                                                     
2009 1   28.2561  17.0188   7.75   0   0   0   0         NaN          NaN   
2010 1   28.2730  17.0161   8.36   0   0   0   0         NaN          NaN   
2011 1   28.2801  17.0169   8.36   0   0   0   0         NaN          NaN   
2012 1   28.2498  17.0395  10.65   0   0   0   0         NaN          NaN   
2013 1   28.2313  17.0458  12.15   0   0   0   0         NaN          NaN   
2014 1   28.2312  17.0616  12.68   0   0   0   0         NaN          NaN   
2015 1   28.2390  17.0602  11.90   0   0   0   0         NaN          NaN   
2016 1   28.2518  17.0710  11.69   0   0   0   0         NaN          NaN   
2017 1   28.2684  17.0776  11.21   0   0   0   0         NaN          NaN   
2018 1   28.2776  17.0799  10.61   0   0   0   0         NaN          NaN   
2019 1   28.2824  17.0700   9.95   0   0   0   0         NaN          NaN   
2009 2   28.5855  17.2839   7.54   1   0   0   0         NaN          NaN   
2010 2   28.6095  17.2886   7.79   1   0   0   0         NaN          NaN   
2011 2   28.6201  17.2961   8.04   1   0   0   0         NaN          NaN   
2012 2   28.6345  17.3053   7.88   1   0   0   0         NaN          NaN   
2013 2   28.6525  17.3139   7.52   1   0   0   0         NaN          NaN   
2014 2   28.6840  17.3220   6.11   1   0   0   0         NaN          NaN   
2015 2   28.7077  17.3305   5.30   1   0   0   0         NaN          NaN   
2016 2   28.7291  17.3401   4.81   1   0   0   0         NaN          NaN   
2017 2   28.7532  17.3457   4.33   1   0   0   0         NaN          NaN   
2018 2   28.7701  17.3540   4.00   1   0   0   0         NaN          NaN   
2019 2   28.7861  17.3621   3.74   1   0   0   0         NaN          NaN   
2009 3   27.9367  16.7487   8.34   0   1   0   0         NaN          NaN   
2010 3   27.9671  16.7592   8.06   0   1   0   0         NaN          NaN   
2011 3   27.9981  16.7676   7.51   0   1   0   0         NaN          NaN   
2012 3   28.0156  16.7765   7.29   0   1   0   0         NaN          NaN   
2013 3   28.0386  16.7871   7.07   0   1   0   0         NaN          NaN   
2014 3   28.0669  16.7904   6.91   0   1   0   0         NaN          NaN   
2015 3   28.0735  16.7956   6.91   0   1   0   0         NaN          NaN   
2016 3   28.0834  16.8051   7.00   0   1   0   0         NaN          NaN   
2017 3   28.1134  16.8185   6.34   0   1   0   0         NaN          NaN   
2018 3   28.1408  16.8283   5.83   0   1   0   0         NaN          NaN   
2019 3   28.1594  16.8485   5.66   0   1   0   0         NaN          NaN   
2009 4   30.4005  18.8742   9.25   0   0   1   0         NaN          NaN   
2010 4   30.4273  18.8740   9.63   0   0   1   0         NaN          NaN   
2011 4   30.4426  18.8746   8.95   0   0   1   0         NaN          NaN   
2012 4   30.4652  18.8810   8.07   0   0   1   0         NaN          NaN   
2013 4   30.4834  18.8832   7.37   0   0   1   0         NaN          NaN   
2014 4   30.5061  18.8878   6.17   0   0   1   0         NaN          NaN   
2015 4   30.5328  18.8947   5.28   0   0   1   0         NaN          NaN   
2016 4   30.5493  18.9059   4.87   0   0   1   0         NaN          NaN   
2017 4   30.5715  18.9152   4.36   0   0   1   0         NaN          NaN   
2018 4   30.6005  18.9233   3.90   0   0   1   0         NaN          NaN   
2019 4   30.6232  18.9341   3.67   0   0   1   0         NaN          NaN   
2009 5   28.7176  17.5527   7.74   0   0   0   1         NaN          NaN   
2010 5   28.7585  17.5520   6.97   0   0   0   1         NaN          NaN   
2011 5   28.7970  17.5467   5.82   0   0   0   1         NaN          NaN   
2012 5   28.8012  17.5497   5.38   0   0   0   1         NaN          NaN   
2013 5   28.8056  17.5582   5.23   0   0   0   1         NaN          NaN   
2014 5   28.8274  17.5640   4.98   0   0   0   1         NaN          NaN   
2015 5   28

In [10]:
y = df['y']
X = df[['x1','x2']]

# Pooled

In [13]:
import linearmodels
from linearmodels.panel import PooledOLS
import statsmodels.api as sm

exog_vars = ["x1","x2"]
exog = sm.add_constant(df[exog_vars])
pooled_mod = PooledOLS(df.y, exog)
pooled_res = pooled_mod.fit(cov_type = 'clustered',
                    cluster_entity = True)
print(pooled_res)

                          PooledOLS Estimation Summary                          
Dep. Variable:                      y   R-squared:                        0.9945
Estimator:                  PooledOLS   R-squared (Between):              0.9156
No. Observations:                  55   R-squared (Within):               0.9948
Date:                Wed, Jun 07 2023   R-squared (Overall):              0.9945
Time:                        23:26:38   Log-likelihood                    72.720
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      4700.1
Entities:                          11   P-value                           0.0000
Avg Obs:                       5.0000   Distribution:                    F(2,52)
Min Obs:                       5.0000                                           
Max Obs:                       5.0000   F-statistic (robust):          2.265e+05
                            

In [16]:
reg= sm.OLS(df.y, exog).fit(cov_type="HC3")
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.994
Model:                            OLS   Adj. R-squared:                  0.994
Method:                 Least Squares   F-statistic:                 1.053e+04
Date:                Wed, 07 Jun 2023   Prob (F-statistic):           1.51e-68
Time:                        23:27:25   Log-Likelihood:                 72.720
No. Observations:                  55   AIC:                            -139.4
Df Residuals:                      52   BIC:                            -133.4
Df Model:                           2                                         
Covariance Type:                  HC3                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          8.3982      0.163     51.377      0.000       8.078       8.719
x1             1.1726      0.009    136.257      0.000       1.156       1.189
x2            -0.0118      0.004     -3.367      0.001      -0.019      -0.005
==============================================================================
Omnibus:                        8.091   Durbin-Watson:                   0.346
Prob(Omnibus):                  0.017   Jarque-Bera (JB):                8.518
Skew:                          -0.959   Prob(JB):                       0.0141
Kurtosis:                       2.808   Cond. No.                         488.
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC3)
"""

# Fixed

In [25]:
from linearmodels.panel import PanelOLS
FE_mod = PanelOLS(df.y, exog, entity_effects = True, time_effects = True)
FE_res = FE_mod.fit(cov_type = 'clustered',
                    cluster_entity = True)
print(FE_res.summary)

                          PanelOLS Estimation Summary                           
Dep. Variable:                      y   R-squared:                        0.8982
Estimator:                   PanelOLS   R-squared (Between):              0.9610
No. Observations:                  55   R-squared (Within):               0.8933
Date:                Wed, Jun 07 2023   R-squared (Overall):              0.8936
Time:                        23:32:34   Log-likelihood                    184.93
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      167.58
Entities:                          11   P-value                           0.0000
Avg Obs:                       5.0000   Distribution:                    F(2,38)
Min Obs:                       5.0000                                           
Max Obs:                       5.0000   F-statistic (robust):             520.38
                            

# Random

In [18]:
from linearmodels.panel import RandomEffects
RE_mod = RandomEffects(df.y, exog)
RE_res = RE_mod.fit(cov_type = 'clustered',
                    cluster_entity = True)
print(RE_res.summary)

                        RandomEffects Estimation Summary                        
Dep. Variable:                      y   R-squared:                        0.9945
Estimator:              RandomEffects   R-squared (Between):              0.9156
No. Observations:                  55   R-squared (Within):               0.9948
Date:                Wed, Jun 07 2023   R-squared (Overall):              0.9945
Time:                        23:29:44   Log-likelihood                    72.720
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      4700.1
Entities:                          11   P-value                           0.0000
Avg Obs:                       5.0000   Distribution:                    F(2,52)
Min Obs:                       5.0000                                           
Max Obs:                       5.0000   F-statistic (robust):          2.265e+05
                            

# Fixed Vs. Random

# тест Хаусмана

# Тест Бреуша-Пагана

In [31]:
from linearmodels.panel import RandomEffects
from linearmodels.panel import compare
from statsmodels.stats.diagnostic import het_breuschpagan
# Создание модели панельных данных
model = PanelOLS(df.y, exog, entity_effects=True, time_effects=False)
# Оценка модели
results = model.fit()
# Получение остатков
residuals = results.resids

# Выполнение теста Бройша-Пагана
bp_test = het_breuschpagan(residuals, exog)

print("Breusch-Pagan Test Statistic:", bp_test[0])
print("P-value:", bp_test[1])

Breusch-Pagan Test Statistic: 12.328165730387987
P-value: 0.0021036468078402507


Исходя из результатов теста Бреуша-Пагана, делаем выбор в пользу модели со случайными эффектами. Так как p_value < 0.05, то делаем выбор в пользу модели со случайными эффектами.

# Тест Чоу

In [ ]:
exog_vars = ["x1","x2"]
exog = sm.add_constant(df[exog_vars])

In [32]:
df_for_chow = df.copy()
df_for_chow['treat'] = 0
df_for_chow['treat'][int(len(df['x1'])/2):] = 1

# df_for_chow=df_for_chow.sort_values(by='treat', ascending=True).reset_index().drop(columns=['id'])

break_point = int(len(df['x1'])/2)


def linear_residuals(X, y):
    import pandas as pd
    import numpy as np
    from sklearn.linear_model import LinearRegression as lr

    #строим линейную модель
    model = lr().fit(X, y)

    # строим датафрейм с предсказанными значениями целевой метрики 
    summary_result = pd.DataFrame(columns = ['y_hat'])
    yhat_list = model.predict(X)
    summary_result['y_hat'] = yhat_list  

    # добавляем к датафрейму реальные значения целевой метрики
    summary_result['y_actual'] = y.values

    # вычисляем остатки
    summary_result['residuals'] = summary_result.y_actual - summary_result.y_hat

    # возводим остатки в квадрат
    summary_result['residuals_sq'] = summary_result.residuals ** 2
    return(summary_result)

# пишем функцию, которая считает сумму квадратов остатков 
def calculate_RSS(X, y):
    resid_data = linear_residuals(X, y)
    rss = resid_data.residuals_sq.sum()
    return(rss)

def ChowTest(X, y, last_index_in_model_1, first_index_in_model_2):
    rss_pooled = calculate_RSS(X, y)
    # делим выборку на две подвыборке, в которой есть “слом”, наличие которого мы хотим проверить, и в котором слома нет

    X1 = X.iloc[:last_index_in_model_1]
    y1 = y.iloc[:last_index_in_model_1]
    rss1 = calculate_RSS(X1, y1)

    X2 = X.iloc[first_index_in_model_2:]
    y2 = y.iloc[first_index_in_model_2:]
    rss2 = calculate_RSS(X2, y2)

    # находим кол-во регрессоров + 1 для константы
    k = X.shape[1] + 1

    # находим кол-во наблюдений до слома
    N1 = X1.shape[0]

    # находим кол-во наблюдений после слома
    N2 = X2.shape[0]

    # вычисляем числитель для статистики Чоу
    numerator = (rss_pooled - (rss1 + rss2)) / k

    # вычисляем знаменатель для статистики Чоу
    denominator = (rss1 + rss2) / (N1 + N2 - 2 * k)

    # вычисляем статистику Чоу
    Chow_Stat = numerator / denominator
    
    # статистика Чоу имеет распределение Фишера с k и N1 + N2 - 2k степенями свободы
    from scipy.stats import f
    
    # считаю p-value
    p_value = 1 - f.cdf(Chow_Stat, dfn = 5, dfd = (N1 + N2 - 2 * k))
    result = (Chow_Stat, p_value)
    print('Chow_Stat', Chow_Stat)
    print('p_value', p_value)
    return(result)

ChowTest(y=df.y, X = exog,  
        last_index_in_model_1=break_point-1,  
        first_index_in_model_2=break_point)[1]

<ipython-input-32-59d7d3977aee>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_for_chow['treat'][int(len(df['x1'])/2):] = 1


Chow_Stat 3.4606115775852015
p_value 0.00974923731698929


0.00974923731698929

# P-value теста Чоу равно 0.009749 <0.05. Между pooled OLS и FE выбираем модель с фиксированными эффектами.

# Предпочтительная модель - с случайными эффектами